File: Java/Beginner/g_Tree/boj_5052.java
package Beginner.g_Tree;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

public class boj_5052 {
    static int t;
    static int n;

    static class TrieNode {
        Map<Character, TrieNode> childNodes = new HashMap<>();
        boolean isFinal = false;
    }

    static class Trie {
        TrieNode root = new TrieNode();

        boolean insert(String number) {
            TrieNode curNode = root;

            for (int i = 0; i < number.length(); i++) {
                char ch = number.charAt(i);

                if (curNode.childNodes.get(ch) == null) {
                    curNode.childNodes.put(ch, new TrieNode());
                }

                curNode = curNode.childNodes.get(ch);

                if (curNode.isFinal) {
                    return false;
                }
            }

            if (curNode.childNodes.size() != 0) {
                return false;
            }

            curNode.isFinal = true;
            return true;
        }

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            n = Integer.parseInt(br.readLine());

            boolean flag = true;

            Trie trie = new Trie();

            for (int i = 0; i < n; i++) {
                if (!trie.insert(br.readLine())) {
                    flag = false;
                }
            }

            if (flag) {
                sb.append("YES").append("\n");
            } else {
                sb.append("NO").append("\n");
            }
        }

        System.out.println(sb);
    }
}

File: v1/Kotlin/src/main/kotlin/boj/beginner/g_Tree/boj_5052.kt
package boj.beginner.g_Tree

import java.lang.NullPointerException

private class Trie {
    class Node {
        val childNodes = HashMap<Char, Node>()
        var isFinal = false
    }

    val root = Node()

    fun insert(numbers: String): Boolean {
        var curNode = root

        numbers.forEach {
            if (curNode.childNodes[it] == null) curNode.childNodes[it] = Node()
            curNode = curNode.childNodes[it] ?: throw NullPointerException()
            if (curNode.isFinal) return false
        }

        if (curNode.childNodes.size != 0) return false
        curNode.isFinal = true
        return true
    }
}

fun boj_5052() = with(System.`in`.bufferedReader()) {
    val t = readLine().toInt()
    val sb = StringBuilder()

    repeat(t) {
        val n = readLine().toInt()
        val trie = Trie()
        var flag = true

        repeat(n) {
            if (!trie.insert(readLine())) flag = false
        }

        if (flag) sb.append("YES")
        else sb.append("NO")
        sb.append("\n")
    }

    print(sb)
}

fun boj_5052_2() = with(System.`in`.bufferedReader()) {

    fun checkContainsString(nums: List<String>): Boolean {
        for (i in 0 until nums.size - 1)
            if (nums[i + 1].startsWith(nums[i])) return false

        return true
    }

    val t = readLine().toInt()
    val sb = StringBuilder()

    repeat(t) {
        val n = readLine().toInt()
        val phoneNums = ArrayList<String>()
        var flag = true

        repeat(n) { phoneNums.add(readLine()) }

        if (!checkContainsString(phoneNums.sorted())) flag = false

        if (flag) sb.append("YES")
        else sb.append("NO")
        sb.append("\n")
    }

    print(sb)
}